Implementation of scan

bohnstingl opened the pull request:
This operation is supposed to be the pendant to the `associative_scan`, but can operate with non-associative functions. 

@ydwu4 


cc @albanD @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @rec

bohnstingl committed:
First implementation of scan

bohnstingl committed:
Added RNN test

bohnstingl committed:
WIP: fixed typo and added scan decomposition

bohnstingl committed:
Fixed shape checks scan operator

bohnstingl committed:
Advanced scan operation

bohnstingl committed:
Used backend=eager in scan
Reverted while_loop for scan

bohnstingl committed:
Updated compile_cnt testcase with reverse flag

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into scan_operator

bohnstingl committed:
WIP: Scan after merge with main

bohnstingl committed:
Fixed lint issues and skipped scan tests that require inductor

bohnstingl committed:
Removed concatenate operations from generic_scan

bohnstingl committed:
Removed unnecessary reverse flag from dispatch keys

bohnstingl committed:
Fixed lintrunner and graph test issue

bohnstingl committed:
Extended scan with init parameter to be consistent with jax.lax.scan

ydwu4 commented on test/functorch/test_control_flow.py:
```diff
@@ -1314,6 +1396,220 @@ def mul(x: torch.Tensor, y: torch.Tensor):
+    @parametrize("reverse", [False, True])
+    @parametrize("device", [torch.device("cpu"), torch.device("cuda")])
+    def test_scan_compile_cnt(self, reverse, device):
+        def add(x: torch.Tensor, y: torch.Tensor):
```
Do we only support combine functions that takes two tensor inputs? Can the combine_fn support pytrees?
```
def combine(x_dict, y_dict):
  ...
```

bohnstingl replied:
No, like in the case of `associative_scan`, the combine_fn supports pytrees as inputs. I have added testcases for that.

ydwu4 commented on test/functorch/test_control_flow.py:
```diff
@@ -1238,11 +1258,57 @@ def mul(x: torch.Tensor, y: torch.Tensor):
         self.assertEqual(cumsum1, cumsum_exp)
 
+    def test_pointwise_scan_simple_graph(self):
+        def add(x: torch.Tensor, y: torch.Tensor):
```
Do we support combine functions that takes pytree inputs? e.g.
```
def combine(a_dict_list, b_dict_list):
  ...
```
If not we should state this clearly in the linline doc of scan.

bohnstingl replied:
No, we support pytree inputs as well. Let me know if this is not well enough stated in the docstring of scan.

ydwu4 commented on test/functorch/test_control_flow.py:
```diff
@@ -1314,6 +1396,220 @@ def mul(x: torch.Tensor, y: torch.Tensor):
+        rnn.load_state_dict(new_state_dict)
+
+        def RNN(x: torch.Tensor, y: torch.Tensor):
+            W_ih = torch.ones((5, 7), device=device)
```
We need to support closures/nn_modules in order for an actual RNN model to work. Please left a TODO for supporting closures.

bohnstingl replied:
I have added a TODO for this in scan. This test was just the first attempt to realize an RNN and it somewhat works.

ydwu4 commented on test/functorch/test_control_flow.py:
```diff
@@ -1314,6 +1396,220 @@ def mul(x: torch.Tensor, y: torch.Tensor):
+            self.assertEqual(cnt.frame_count, 1)
+
+            x = torch.randn(3, 20, 5, device=device)
+            # Recompilation
```
Please add an explantion for why there is compilation/no recompilation.

bohnstingl replied:
I added a description of what I think should happen and why recompilations, or no recompilations are done.

ydwu4 commented on test/functorch/test_control_flow.py:
```diff
@@ -1314,6 +1396,220 @@ def mul(x: torch.Tensor, y: torch.Tensor):
+        inp = torch._ops.ops.aten.slice(x, dim, 1, None, 1)
+        # Init scalar
+        init = 1.0
+        with self.assertRaisesRegex(Exception, ".*"):
```
Testing exceptions should be in another test. Also please don't just catch all Exception (should catch specific action with proper error message). Otherwise, it could cause silent errors.

bohnstingl replied:
Right, I fully agree that catching generic Exceptions is bad, but I couldn't really figure out a way how to do that properly. The problem is that the AssertionError are masked by some errors coming from dynamo, e.g., 

```
torch._dynamo.exc.Unsupported: should_compile_partial_graph=False

from user code:
   File "/data_malta3_ssd/pytorch/torch/_higher_order_ops/scan.py", line 158, in scan
    assert all(
```

Do you have any suggestions how to properly do this?

ydwu4 replied:
Can you show me the full error message? I guess that's because of a graph break. This is because dynamo will inline the code in scan.py and somewhere in `assert all(...)` is causing a graph break. Because we torch.compile(fullgraph=True) hence get this error. 


We could either fix the graph break in dynamo (if we can) or try to use an supported approach to do the same thing (e.g. can you try raise an runtime error or something?)

bohnstingl replied:
I see. For the `cond.py` we had changed the `asserts` into `RuntimeError` which where supported and hence could be properly catched in the testcase. I did the same here and also aligned the respective cases in `associative_scan`. As discussed, if dynamo breaks the graph due to error, the `assertRaisesRegex` will catch `torch._dynamo.exc.Unsupported, "Observed exception.*"` for now.

ydwu4 commented on test/functorch/test_control_flow.py:
```diff
@@ -1314,6 +1396,220 @@ def mul(x: torch.Tensor, y: torch.Tensor):
+    @unittest.skipIf(not SM70OrLater, "triton")
+    @unittest.skipIf(not torch.cuda.is_available(), "Test requires CUDA.")
+    @parametrize("reverse", [False, True])
+    @parametrize("device", [torch.device("cpu"), torch.device("cuda")])
```
Why we have GPU tests? Since we don's support device = gpu right. Please also fix other tests.

bohnstingl replied:
I am not sure about this one here. We do not support `_inductor`, but our implementation in `generic_scan` does support CUDA devices. We have a lot of testcases that demonstrate this.

ydwu4 replied:
oh, i'm mainly referring to `@unittest.skipIf(not SM70OrLater, "triton")`, because triton is required only when we're doing codegen in inductor.

ydwu4 replied:
Discussed offline, we'll use @requires_cuda

bohnstingl replied:
As per our discussion I replaced `@unittest.skipIf(not torch.cuda.is_available(), "Test requires CUDA.")` with `from torch.testing._internal.common_utils import requires_cuda`

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -0,0 +1,349 @@
+        num_elems = elems[0].shape[dim]
+
+        ind = 0
+        xs = elems_init
```
The naming here is confusing to me lol. We should just rename xs -> inits because the xs in jax's scan actually means elems.

bohnstingl replied:
I tried to use a better naming now. In particular, I used `input` (to be consistent with associative_scan) and `init` (to be more similar to jax.lax.scan).

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -0,0 +1,349 @@
+        )
+
+        # Approach with concatenate
+        # outs = list(safe_map(cmb, outs, xs))
```
let's keep the without concatenation approach and write a proper comment for what's each step is doing.

bohnstingl replied:
Done

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -0,0 +1,349 @@
+        def add(x: torch.Tensor, y: torch.Tensor):
+            return x + y
+
+        cumsum = scan(add, x, dim)
```
Give an examplanation for the case where init is None. and give an example similar to (please make it look nicer).
```
scan(f, (a, b, c)) = (a, f(a, b), f(f(a ,b), c))
```

bohnstingl replied:
Done

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -0,0 +1,349 @@
+                combine_fn, input, dim=dim, reverse=reverse, init=init
+            )
+
+    leaves_init, spec_init = pytree.tree_flatten(init)
```
1. we didn't have tests for pytree init values i think
2. torch.compile should be able to flatten the init and input, so here, we can expect them to be plain tuples.

bohnstingl replied:
1. I added a testcase for pytree init values

2. I am not entirely sure what you mean? The `pytree.tree_flatten` is the same as we do for the `input`? I thought that this is the place where the pytree from the user is flattened into a plain tuple and the respective shape?

ydwu4 replied:
I did a double check. Yeah, for scan, the pytrees are flattened in the wrapper code before entering dynamo. This is the opposite of cond or while_loop, where pytrees are flattened in dynamo. This is not ideal but I'm OK with the current status. Can you also add a `TODO: unify handing of pytrees for control flow ops`?

bohnstingl replied:
This handling is the same as in `associative_scan`, if you prefer the way how cond or while_loop does it, we can switch it in a follow-up PR? 
For the moment, I added a [TODO](https://github.com/pytorch/pytorch/blob/810491fddfb4e51cdbe4e56c7094c8a56729c5bd/torch/_higher_order_ops/scan.py#L97) to `scan`

ydwu4 replied:
This is OK.

ydwu4 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -1106,6 +1108,153 @@ def arg_extractor(combine_fn, input, dim):
+                args=(
+                    SourcelessBuilder.create(
+                        tx,
+                        leaf_init.size
```
why they're called leaf? It would be better to keep consistent with `elems` `inits` or `xs`, "init" in jax style across all places.

bohnstingl replied:
I tried to rename everything. Now it is `input` (to be consistent with the associative_scan) and `init` as it is called in jax.lax.scan

ydwu4 reviewed (commented):
Please take a look at the comments. Overall it's a bit overwhelming for me to review the diff. Let's clean it up according to the comments and let me know if need another round of reviewing.

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into scan_operator

bohnstingl committed:
Incorporated suggested changes from review

bohnstingl committed:
Updated higher_order_ops.py

bohnstingl commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -1106,6 +1108,153 @@ def arg_extractor(combine_fn, input, dim):
+
+        combine_fn, input, init, dim = arg_extractor(*args, **kwargs)
+
+        if input.python_type() != list:
```
I am not sure if this TODO is needed? When this point is reached, the `input` and `init` should already be flattened to a list, correct? This check is also present for the associative_scan and I just overtook it.

ydwu4 replied:
This is before the flattening i believe. The actual flattening is done by speculate_subgraph. Can you add a test to make sure?

ydwu4 replied:
I was wrong, the input here is already flattened.

bohnstingl replied:
Yes, we have testcases where the input, the init and both are pytrees. In all cases, they are flattened at this point.

bohnstingl commented on torch/_higher_order_ops/scan.py:
```diff
@@ -0,0 +1,349 @@
+    input: pytree.PyTree,
+    dim: int,
+    reverse: bool = False,
+    init: pytree.PyTree = [],  # noqa: B006
```
Done

bohnstingl commented on torch/_higher_order_ops/scan.py:
```diff
@@ -0,0 +1,349 @@
+    init: pytree.PyTree = [],  # noqa: B006
+) -> torch.Tensor:
+    r"""
+    Performs an inclusive scan with an associative pointwise combine function.
```
I reviewed the docstring of `scan` and also of `associative_scan` and adjusted it a bit

bohnstingl commented on torch/_higher_order_ops/scan.py:
```diff
@@ -0,0 +1,349 @@
+        input (torch.Tensor): The input tensor, or nested pytree of tensors.
+        dim (int): the dimension to scan over
+        reverse (bool): A boolean stating if the scan should be reversed with respect to the dimension.
+        init (torch.Tensor): The inital scan carry, a tensor, or nested pytree of tensors.
```
I reviewed the docstring of `scan` and also of `associative_scan` and adjusted it a bit

bohnstingl commented on torch/_higher_order_ops/scan.py:
```diff
@@ -0,0 +1,349 @@
+    Args:
+        combine_fn (Callable): A binary callable with type ``(Tensor, Tensor) -> Tensor``,
+            or if input is a pytree ``(pytree, pytree) -> pytree``.
+            This function must be pure, pointwise, and satisfy the associative property.
```
I reviewed the docstring of `scan` and also of `associative_scan` and adjusted it a bit

bohnstingl commented:
@ydwu4, I incorporated your suggestions and updated the code.
I also merge it with the latest main, which includes the associative_scan feature. So there have been a lot of new test cases added and in all of them, I added the `scan` as well.

I would be more than happy about another round of review if you can. Thank you verymuch!

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into scan_operator

bohnstingl committed:
Removed unnecessary code

ydwu4 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -1117,6 +1119,153 @@ def arg_extractor(combine_fn, input, dim):
+            combine_fn,
+            sub_args,
+            sub_kwargs={},
+            description="scan_combine",
```
De-dupe with associtiave_scan's description? Probably `associativ_scan_combine_fn` vs `scan_combine_fn`

bohnstingl replied:
Done

ydwu4 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -1117,6 +1119,153 @@ def arg_extractor(combine_fn, input, dim):
+                )
+
+        combine_gm = torch.fx.GraphModule(dict(tx.output.nn_modules), combine_graph)
+        combine_fn_name = add_subgraph(tx, "scan_combine", combine_gm)
```
De-dup `associatie_scan_combine_fn` and `scan_combine_fn`?

bohnstingl replied:
Done

ydwu4 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -1117,6 +1119,153 @@ def arg_extractor(combine_fn, input, dim):
+
+        with tx.fake_mode:
+            out_meta = tuple(
+                ini_proxy.node.meta["example_value"].clone() for ini_proxy in init_proxy
```
This is probably wrong? The example output of the scan operator should be the the inits stacked together, right?  We have similar logic for map in https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/variables/higher_order_ops.py#L1185-L1196

To test this, we can have a downstream operator that takes the output of scan as output e.g.
suppose scan() -> (3, 2), we call Tensor (3, 3)  @ (3, 2).

bohnstingl replied:
Good catch!
The code here would still work as init necessarily has a `1` at the dim of scan and then later automatic broadcasting should take care of  properly expanding it. However, I agree with you that we should have a correct version and implemented one taking inspiration from map.
I also added a testcase, [test_pointwise_scan_downstream_op](https://github.com/pytorch/pytorch/blob/810491fddfb4e51cdbe4e56c7094c8a56729c5bd/test/functorch/test_control_flow.py#L1555)

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -0,0 +1,339 @@
+        # Usage without the usage of ``init``
+        cumsum = scan(add, x, dim)
+        # This produces the output
+        cumsum = [x0, add(x0, x1), add(x1, x2)]
```
1. The output dtype should be a tensor, right? We can have torch.stack around the current output list to make this clearer.
2. the output should look like [x0, add(x0, x1), add(add(x0, x1), x2)], right? Right now, it's not a accumulative sum i think.
3. When init is not None, is init included as part of output? What's the shape of the output in the scanned dimension?

bohnstingl replied:
Ad 1.) I corrected the docstring
Ad 2.) You are fully right, I corrected the docstring
Ad 3.) Init is included as part of the output. I tried to cover this in the example in the docstring. In the current implementation, the shape of init has to be the same shape as the output of the scan combine_fn. The shape of output for the scanned dimension is always 1.

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -0,0 +1,339 @@
+    unwrapped_input = ctx.unwrap_tensors(input)
+    unwrapped_init = ctx.unwrap_tensors(init)
+    with ctx.redispatch_to_next() as m:
+        functional_combine_fn = ctx.functionalize(combine_fn)
```
We should also check combine_fn shouldn't have input aliasing or mutations. Similar logic can be found in cond and while loop's functionalization key implementation.

bohnstingl replied:
I added [testcases](https://github.com/pytorch/pytorch/blob/810491fddfb4e51cdbe4e56c7094c8a56729c5bd/test/functorch/test_control_flow.py#L4484-L4542) for these checks

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -0,0 +1,339 @@
+@scan_op.py_impl(FakeTensorMode)
+def assoiciative_scan_fake_tensor_mode(mode, combine_fn, input, init, dim):
+    with mode:
+        return combine_fn(*input, *input)
```
This is not correct because the fake output of scan should be the stacked output of combine_fn.

bohnstingl replied:
I corrected this part. Interesting there was no testcase failing because of that. I also added a [dedicated test](https://github.com/pytorch/pytorch/blob/810491fddfb4e51cdbe4e56c7094c8a56729c5bd/test/test_fake_tensor.py#L928) for the FakeTensorMode in test_fake_tensor.py.

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -0,0 +1,339 @@
+
+@scan_op.py_impl(ProxyTorchDispatchMode)
+def scan_proxy_mode(mode, combine_fn, input, init, dim):
+    return trace_scan(mode, scan_op, combine_fn, input, init, dim)
```
There're cases that we don't need to trace. Can follow cond/while_loop for the special implementation

bohnstingl replied:
This is obsolete

ydwu4 reviewed (commented):
We're closer to merge!

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -0,0 +1,339 @@
+    return combined_leaves
+
+
+def scan(
```
In fact, we might want to support the case where combined_output is different from carry: i.e.
```python
# This is our current semantic
def combine_fn(init, input):
  # next iteration's init is current iteration's output
   return init + input
```

However, there are cases that the carry is different from output e.g.
```
def combind_fn(init, input):
  # next iteration's input is init + input
  # but current iteration's output is init
  return init + input, init
```
Another jax example would be:
```
import jax.numpy as jnp
from jax import lax

def body(init, x):
    return (init + x, init + x.sum())

xs = jnp.array([[1, 2, 3, 4], [5, 6, 7, 8], [9, 10, 11, 12]])
init = jnp.array([0, 0, 0, 0], dtype=xs.dtype)
result = lax.scan(body, init, xs)
print(result)
```

This is kind of bigger change that we need to do.

ydwu4 replied:
The other implication of this change is that we probably want to split the tests of associative with scan. In the future, we probably want to extend associative_scan to have the same interface of scan instead of the other way around.

bohnstingl replied:
I added a [TODO](https://github.com/pytorch/pytorch/blob/810491fddfb4e51cdbe4e56c7094c8a56729c5bd/torch/_higher_order_ops/scan.py#L98) for this new user interface

ydwu4 replied:
I think we need to do it quickly (best in this pr). I'm also OK with do it in the next pr since the current PR is too large. Autograd PR would require and we really don't want to expose the current interface to users.

bohnstingl committed:
Integrated code review findings

bohnstingl commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -1117,6 +1119,153 @@ def arg_extractor(combine_fn, input, dim):
+
+        # Trace the subgraph
+        # TODO: Fix these pointless new_empty calls appearing in the dynamo output graph.
+        sub_args = [
+            inp.call_method(
+                tx,
+                "new_empty",
+                args=(
+                    SourcelessBuilder.create(
+                        tx,
+                        ini.size
+                        if ini.size is not None
+                        else tuple(
+                            [  # noqa: C416
+                                it
+                                for it in BuiltinVariable(getattr)
+                                .call_function(
+                                    tx,
+                                    [ini, ConstantVariable.create("shape")],
+                                    {},
+                                )
+                                .items
+                            ]
+                        ),
+                    ),
+                ),
+                kwargs={
+                    "dtype": SourcelessBuilder.create(tx, inp.dtype),
+                    "device": SourcelessBuilder.create(tx, inp.device),
+                    "requires_grad": SourcelessBuilder.create(tx, inp.requires_grad),
+                },
+            )
+            for inp, ini in itertools.chain(
+                zip(input.items, init.items), zip(input.items, init.items)
+            )
+        ]
```
I added a [TODO](https://github.com/pytorch/pytorch/blob/810491fddfb4e51cdbe4e56c7094c8a56729c5bd/torch/_dynamo/variables/higher_order_ops.py#L1165) for this.

bohnstingl commented on torch/_higher_order_ops/associative_scan.py:
```diff
@@ -101,12 +101,13 @@ def associative_scan(
         combine_fn (Callable): A binary callable with type ``(Tensor, Tensor) -> Tensor``,
             or if input is a pytree ``(pytree, pytree) -> pytree``.
-            This function must be pure, pointwise, and satisfy the associative property.
+            This function must be pure, i.e., no lifted arguments are supported at the moment,
```
Done

bohnstingl commented on torch/_higher_order_ops/scan.py:
```diff
@@ -0,0 +1,339 @@
+        else:
+            leaves_input = []
+
+    if len(leaves_input) > 0:
```
This makes sense to me, I revised the error handling such that a zero scan dimension is not allowed anymore

ydwu4 replied:
Can we remove the if len(leaves_input) check since it's always non-zero length?

bohnstingl commented on torch/_higher_order_ops/scan.py:
```diff
@@ -0,0 +1,339 @@
+        "call_function", func_overload, proxy_args, {}, name="scan"
+    )
+
+    with disable_proxy_modes_tracing():
```
I think a similar reasoning applies here as for the `higher_order_ops.py`. However, I corrected the sample shape computation according to `map.py`.
This is covered in the testcase test_pointwise_scan_downstream_op

bohnstingl commented on test/functorch/test_control_flow.py:
```diff
@@ -1215,32 +1217,42 @@ def fwbw(map_op, f, x, y):
             return x * y
 
+        def scan_op_wrapper(*args, **kwargs):
+            if scan_op == scan and "combine_mode" in kwargs:
```
Done

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -0,0 +1,377 @@
+
+    if len(leaves_init) == 0 and len(leaves_input) == 0:
+        raise RuntimeError("Expected at least 1 init or input leaf")
+    if len(leaves_init) > 0:
+        if any(not isinstance(x, torch.Tensor) for x in leaves_init):
+            raise RuntimeError("Init leaves must be a Tensor")
+        if any(x.shape[dim] == 0 for x in leaves_init):
+            raise RuntimeError("Init leaves must have a scan dimension > 0")
+    if len(leaves_input) > 0:
+        if any(not isinstance(x, torch.Tensor) for x in leaves_input):
+            raise RuntimeError("Input leaves must be a Tensor")
+        if any(x.shape[dim] == 0 for x in leaves_input):
+            raise RuntimeError("Input leaves must have a scan dimension > 0")
```
Redundant code?

ydwu4 commented:
Waiting for CI. Also really appreciate it if we can change the interface in this PR (also the splitting the tests). Yeah, it's a lot of work but it will make the tests and code look cleaner.

bohnstingl committed:
Revised user interface
Reworked testcases (separated associative_scan from scan)
Added additional testcases for scan
Integrated code review comments

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into scan_operator

bohnstingl committed:
Fixed linterunner issues and remove unnecessary testcases

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -0,0 +1,402 @@
+        if any(x.shape[dim] == 0 for x in arg):
+            raise RuntimeError("All leaves must have a scan dimension > 0")
+
+    check_arg(leaves_init)
```
Leaves of init don't necessarily have the same shape[dim], right? only input is required to have a > 0 dim. 

Do we have a test case for this?

bohnstingl replied:
Correct, init does not have to have the same shape as input and also does not have to have shape[dim] > 0. I have remove this check for the init and also checked that we have a testcase for that. We had a [testcase](https://github.com/pytorch/pytorch/blob/cb06c9cf487b748caa807bcc66207ce2f57f55cf/test/functorch/test_control_flow.py#L2435-L2533) for that already, but I added an additional one.

ydwu4 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -1117,6 +1119,207 @@ def arg_extractor(combine_fn, input, dim):
+        combine_carry_proxy = combine_result.items[0].as_proxy()
+        combine_output_proxy = combine_result.items[1].as_proxy()
+        for carry, out, inp_proxy, ini_proxy, s_args in zip(
+            combine_carry_proxy, combine_output_proxy, input_proxy, init_proxy, sub_args
```
combine_carry_proxy and combine_output_proxy doesn't have to have same length right? because they're different pytree structures. I think we might want to split checks according to same pytree length.

bohnstingl replied:
I split the checks into checks on init and carry and then checks on input

ydwu4 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -1117,6 +1119,207 @@ def arg_extractor(combine_fn, input, dim):
+            # We also do a clone with contiguous_format. This is to be consistent with
+            # eager semantic of map, which stacks the outputs. The result is contiguous
+            # as a result of the stack operation.
+            fake_out_shapes = [
+                tuple(
+                    -1 if i != dim_fake else dim_len
+                    for i, sh in enumerate(get_fake_value(o.as_proxy().node, tx).size())
+                )
+                for o in combine_result.items[1].items
+            ]
+            out_meta = (
+                [init_p.node.meta["example_value"].clone() for init_p in init_proxy],
+                list(
+                    t.as_proxy()
+                    .node.meta["example_value"]
+                    .expand(*sh)
+                    .clone(memory_format=torch.contiguous_format)
+                    for t, sh in zip(combine_result.items[1].items, fake_out_shapes)
+                ),
+            )
```
Similar logic appears 3 times i think. Probably we should create a helper function for it.

bohnstingl replied:
I consolidated all usecases and created a helper function [make_expanded_output_shape](https://github.com/pytorch/pytorch/blob/cb06c9cf487b748caa807bcc66207ce2f57f55cf/torch/_higher_order_ops/scan.py#L271-L273) for that in `scan.py`

ydwu4 commented on test/functorch/test_control_flow.py:
```diff
@@ -1251,8 +1309,163 @@ def mul(x: torch.Tensor, y: torch.Tensor):
+    @parametrize("compile_mode", ["none", "eager", "compile", "compile_dynamic_shape"])
+    @parametrize("device", [torch.device("cpu"), torch.device("cuda")])
+    # Skipping the scan tests, as scan currently does not support inductor
+    @decorateIf(
+        unittest.skip,
+        lambda params: params["compile_mode"] != "eager"
+        and params["compile_mode"] != "none",
```
Instead of skipping, maybe we should just remove "compile", "compile_dynamic_shape" from compile_mode list? Same comments for the rest of the tests.

bohnstingl replied:
Yeah, my thought on this was that we keep the compilation modes and skip it for the moment, because we will hopefully soon provide an implementation for that and then we can just simply remove the decorator. However, I removed it for the moment to have a cleaner testfile. I left the if-clause though with a `TODO` in the test as a reminder.

ydwu4 commented on test/functorch/test_control_flow.py:
```diff
@@ -1251,8 +1309,163 @@ def mul(x: torch.Tensor, y: torch.Tensor):
+            scan_fct = scan
+
+        # Check all outputs and carries on the correct device and with torch.float32
+        x = torch.randn(3, 10, 2, device=device)
+        op, init = (add, torch.zeros(1, 10, 2, device=device))
+        result = scan_fct(op, init=init, input=x, dim=0, reverse=reverse)
+        result_exp = _fake_scan(op, init=init, input=x, dim=0, reverse=reverse)
+        self.assertEqual(result, result_exp)
+        self.assertEqual(
+            [[r.device.type for r in res] for res in result],
+            [[device.type for _ in res] for res in result],
+        )
+        self.assertEqual(
+            [[r.dtype for r in res] for res in result],
+            [[torch.float32 for _ in res] for res in result],
+        )
+
+        # Check all outputs and carries on the correct device and
+        # carry.dtype torch.float32 and output.dtype torch.float16
+        x = torch.randn(3, 10, 2, device=device, dtype=torch.float16)
+        op, init = (add, torch.zeros(1, 10, 2, device=device, dtype=torch.float32))
+        result = scan_fct(op, init=init, input=x, dim=0, reverse=reverse)
+        result_exp = _fake_scan(op, init=init, input=x, dim=0, reverse=reverse)
+        self.assertEqual(result, result_exp)
+        self.assertEqual(
+            [[r.dtype for r in res] for res in result],
+            [
+                [torch.float32 for _ in range(len(result[0]))],
+                [torch.float16 for _ in range(len(result[1]))],
+            ],
+        )
+
+        # Check all outputs and carries on the correct device and
+        # carry.dtype torch.int64 and output.dtype torch.float32
+        x = torch.randn(3, 10, 2, device=device, dtype=torch.float32)
+        op, init = (add, torch.zeros(1, 10, 2, device=device, dtype=torch.float16))
```
Can we parameterize on dtype like x_dtype, init_dtype? so we don't have list out all the dtypes.

bohnstingl replied:
I parameterized the test with `dtype` and included most of the dtypes that torch supports. See [here](https://github.com/pytorch/pytorch/blob/cb06c9cf487b748caa807bcc66207ce2f57f55cf/test/functorch/test_control_flow.py#L1494-L1549)

ydwu4 commented on test/functorch/test_control_flow.py:
```diff
@@ -1471,44 +1726,860 @@ def fct_pointwise(x, y):
+    @parametrize("reverse", [False, True])
+    @parametrize("device", [torch.device("cpu"), torch.device("cuda")])
+    def test_scan_complex_pytree(self, reverse, device):
+        # Init and input have same pytree
```
init and input doesn't have to have the same pytree. We should remove this restriction.

bohnstingl replied:
Done. The only restrictions now are between the init and the carry. They need to have the same dtype, shape and device

ydwu4 commented on test/functorch/test_control_flow.py:
```diff
@@ -1471,44 +1726,860 @@ def fct_pointwise(x, y):
         torch.compiler.reset()
-        associative_scan1 = torch.compile(associative_scan, fullgraph=True)
-        associative_scan2 = associative_scan
+        if compile_mode == "compile":
```
Worth a helper function.

bohnstingl replied:
Yes, very much so. Done.

ydwu4 commented on test/functorch/test_control_flow.py:
```diff
@@ -1471,44 +1726,860 @@ def fct_pointwise(x, y):
+            )
+            return o2
+
+        torch.compiler.reset()
```
I'm not sure if we need to reset every time.

bohnstingl replied:
No, we don't. Removed.

ydwu4 commented on test/functorch/test_control_flow.py:
```diff
@@ -1471,44 +1726,860 @@ def fct_pointwise(x, y):
         self.assertEqual(result1, expected_result)
-        self.assertEqual(result2, expected_result)
 
     @unittest.skipIf(not SM70OrLater, "triton")
```
We don't need triton for scan tests, right? Please also look at all scan tests to make sure the skip is deleted.

bohnstingl replied:
I have removed the skip triton for the `scan` operator tests.

ydwu4 commented on test/functorch/test_control_flow.py:
```diff
@@ -1471,44 +1726,860 @@ def fct_pointwise(x, y):
+            self.assertEqual(result1, expected_result)
+
+        # Chain with scan
+        def chain_fct_same_dim(inp):
```
Maybe worth separate tests for chained tests.

bohnstingl replied:
Done

ydwu4 commented on test/functorch/test_control_flow.py:
```diff
@@ -1471,44 +1726,860 @@ def fct_pointwise(x, y):
+        unittest.skip,
+        lambda params: (params["device"] == torch.device("cpu")),
+    )
+    def test_generic_scan_non_pointwise(self, reverse, device):
```
The name could be generic_associative_scan_non_pointwise, right? Since we've separated all scan tests with associative_scan tests. Can you help auditing the test names?

bohnstingl replied:
I went through the test names and hopefully changed the appropriately.

ydwu4 commented on test/functorch/test_control_flow.py:
```diff
@@ -1471,44 +1726,860 @@ def fct_pointwise(x, y):
+            RuntimeError,
+            "The size of tensor a.*",
+        ):
+            result_init = scan_fct(op, input=inp, dim=dim, reverse=reverse, init=init)
```
what's going on here?

bohnstingl replied:
This was offline discussed and resolved

ydwu4 commented on test/functorch/test_control_flow.py:
```diff
@@ -1471,44 +1726,860 @@ def fct_pointwise(x, y):
+        self.assertEqual(result_init[0].shape, torch.Size([1, 10, 2]))
+        self.assertEqual(result_init[1].shape, inp.shape)
+
+        # Init wrong pytree
+        init = (
+            torch._ops.ops.aten.slice(x, dim, 0, 1, 1),
+            torch._ops.ops.aten.slice(x, dim, 0, 1, 1),
+        )
+
+        def add_one_carry(x: torch.Tensor, y: torch.Tensor):
+            return x[0], x
+
+        with self.assertRaisesRegex(
+            # Should be: RuntimeError: The number of leaves of the pytree of the new carry produced
+            # by the operator needs to match the length of the pytree of the init
+            torch._dynamo.exc.Unsupported,
+            "Observed exception.*",
+        ):
+            result_init = scan_fct(
+                add_one_carry, input=inp, dim=dim, reverse=reverse, init=init
+            )
```
Mind seperating out all raisesRegex tests into differnt tests (with minimal parameterization)? This would make each unit tests standalone. Mind auditing all the places?

bohnstingl replied:
I separated the raisesRegex out into different test with minimal parameterization as you suggested

ydwu4 commented on test/functorch/test_control_flow.py:
```diff
@@ -1471,44 +1726,860 @@ def fct_pointwise(x, y):
+    @parametrize("reverse", [False, True])
+    @parametrize("device", [torch.device("cpu"), torch.device("cuda")])
+    def test_scan_init_complex_pytree(self, reverse, device):
+        def fct_pointwise_carry_wrong_pytree(x, y):
```
We should probably have a collection of combine_fn and parameterize over them so that they can be shared among tests.

bohnstingl replied:
I added a [collection](https://github.com/pytorch/pytorch/blob/cb06c9cf487b748caa807bcc66207ce2f57f55cf/test/functorch/test_control_flow.py#L167-L223) of such function and consolidated their usecases in the tests

ydwu4 reviewed (commented):
Looking good! Please address the comments.  The comments are mostly related to test organizations and code re-using. Ideally, we want unit tests to be short and standalone (can be parameterized better to cover more cases) and the code to be "dry" (minimal redundancy).

bohnstingl committed:
Incorporated review comments except combine_fn consolidation

bohnstingl committed:
After consolidation of combine_fn in tests

bohnstingl commented on torch/_higher_order_ops/scan.py:
```diff
@@ -0,0 +1,402 @@
+            and the second output  of ``combine_fn`` represents a slice of the output.
+            This function must be pure, i.e., no lifted arguments are supported at the moment
+            and may not have any side effects.
+        init (torch.Tensor): The inital scan carry, a tensor, or nested pytree of tensors.
```
Done

bohnstingl commented on torch/_higher_order_ops/scan.py:
```diff
@@ -0,0 +1,402 @@
+            This function must be pure, i.e., no lifted arguments are supported at the moment
+            and may not have any side effects.
+        init (torch.Tensor): The inital scan carry, a tensor, or nested pytree of tensors.
+            The ``init`` is expected to have the same pytree structure as the first output element
```
Done

bohnstingl commented on torch/_higher_order_ops/scan.py:
```diff
@@ -0,0 +1,402 @@
+        init (torch.Tensor): The inital scan carry, a tensor, or nested pytree of tensors.
+            The ``init`` is expected to have the same pytree structure as the first output element
+            of ``combine_fn``.
+        input (torch.Tensor): The input tensor, or nested pytree of tensors.
```
Done

bohnstingl commented on torch/_higher_order_ops/scan.py:
```diff
@@ -0,0 +1,402 @@
+
+    Example::
+
+        def add(x: torch.Tensor, y: torch.Tensor):
```
Done

bohnstingl commented on torch/_higher_order_ops/scan.py:
```diff
@@ -0,0 +1,402 @@
+    combine_fn: Callable[[pytree.PyTree, pytree.PyTree], pytree.PyTree],
+    init: pytree.PyTree,
+    input: pytree.PyTree,
+    dim: int,
```
I introduced the default dim=0

bohnstingl commented on torch/_higher_order_ops/scan.py:
```diff
@@ -0,0 +1,402 @@
+        def add(x: torch.Tensor, y: torch.Tensor):
+            return x + y, x + y
+
+        i0 = 0.
```
I modified the example slightly and checked that it works indeed as expected

bohnstingl committed:
Fixed lintrunner issues

bohnstingl commented on torch/_higher_order_ops/scan.py:
```diff
@@ -0,0 +1,402 @@
+
+    proxy_mode.tracer.root.register_module(combine_graph_name, combine_graph)
+
+    args = (combine_graph, input, dim)
```
Good catch! Actually we didn't have such a test. I corrected the `trace_scan` and also added a [testcase](https://github.com/pytorch/pytorch/blob/cb06c9cf487b748caa807bcc66207ce2f57f55cf/test/functorch/test_control_flow.py#L2780-L2903) to check that the graph is correct.

bohnstingl committed:
Modified mechanism to handle reverse flag

ydwu4 reviewed (approved):
Looks great. Waiting for CI to pass

bohnstingl committed:
Merge branch 'main' of github.com:pytorch/pytorch into scan_operator

bohnstingl commented:
Thank you @ydwu4 for your reviews and your help on this. I fixed your last comments and merged with the latest `main`.

@albanD, the current PR has been a joint effort with @ydwu4 and went through several iterations of code reviews. It is rather lengthy, but a large portion of that is because we separated `scan` from `associative_scan` on a higher level, which produced quite some code lines. Could you please take a look and maybe approve the PR length as it?

bohnstingl commented on torch/_higher_order_ops/scan.py:
```diff
@@ -0,0 +1,434 @@
+        reverse (bool): A boolean stating if the scan should be reversed with respect to ``dim``, default ``False``.
+
+    Returns:
+        final_carry (torch.Tensor): The final carry of the scan operation
```
This reads good to me.

bohnstingl committed:
Fixed docstring

zou3519 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -0,0 +1,436 @@
+        [pytree.PyTree, pytree.PyTree], Tuple[pytree.PyTree, pytree.PyTree]
+    ],
+    init: pytree.PyTree,
+    input: pytree.PyTree,
```
I would recommend making everything after input kwarg-only

zou3519 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -0,0 +1,436 @@
+    ],
+    init: pytree.PyTree,
+    input: pytree.PyTree,
+    dim: int = 0,
```
Is anyone asking for dim != 0? The API might be simpler if we don't have this initially.

zou3519 replied:
nvm, I see that associative_scan has a dim argument so this makes sense to have

zou3519 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -1065,7 +1067,7 @@ def arg_extractor(combine_fn, input, dim):
             sub_args,
             sub_kwargs={},
-            description="scan_combine",
+            description="associativ_scan_combine_fn",
```
typo

ydwu4 replied:
associative_scan_combine_fn

bohnstingl replied:
Done

zou3519 commented on torch/_dynamo/variables/higher_order_ops.py:
```diff
@@ -1097,7 +1099,7 @@ def arg_extractor(combine_fn, input, dim):
 
         combine_gm = torch.fx.GraphModule(dict(tx.output.nn_modules), combine_graph)
-        combine_fn_name = add_subgraph(tx, "scan_combine", combine_gm)
+        combine_fn_name = add_subgraph(tx, "associatie_scan_combine_fn", combine_gm)
```
typo

ydwu4 replied:
This "associative_scan_combine_fn"

bohnstingl replied:
Done

zou3519 commented on test/functorch/test_control_flow.py:
```diff
@@ -1251,8 +1375,181 @@ def mul(x: torch.Tensor, y: torch.Tensor):
+    @parametrize(
+        "dtype",
+        [
+            torch.float16,
+            torch.float32,
+            torch.float64,
+            torch.int8,
+            torch.int16,
+            torch.int32,
+            torch.int64,
+            torch.complex64,
+            torch.complex128,
```
Why test so much? This feels like overkill

ydwu4 replied:
Yeah, lets keep the cost commonly seen types  e.g. float32, int8, maybe a  torch.complex64?

zou3519 reviewed (changes requested):
from an API perspective:
1) please make everything after `input` kwarg only
2) consider not using "input" as the name of the argument -- input is a special python keyword.

albanD commented:
Waiving PR length check since every line is carefully reviewed and cannot be logically broken down

bohnstingl committed:
Fixed typos

ydwu4 commented:
Talk to Thomas offiline: we'll try to stash the changes to all associative_scan tests and added it back as a follow-up pr.

bohnstingl committed:
Renamed input -> xs for scan and associative_scan
Refactored generic_scan implementation to increase code re-use

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -48,9 +48,8 @@ def scan(
-    input: pytree.PyTree,
-    dim: int = 0,
-    reverse: bool = False,
+    xs: pytree.PyTree,
```
```
def scan(
    combine_fn: Callable[
        [pytree.PyTree, pytree.PyTree], Tuple[pytree.PyTree, pytree.PyTree]
    ],
    init: pytree.PyTree,
    input: pytree.PyTree,
    /,
    dim: int = 0,
    reverse: bool = False,
) -> Tuple[pytree.PyTree, pytree.PyTree]
```
The trick is "/".

bohnstingl replied:
Thank you, I marked combine_fn, init and xs as position-only.

ydwu4 commented on torch/_higher_order_ops/scan.py:
```diff
@@ -92,14 +93,30 @@ def add(x: torch.Tensor, y: torch.Tensor):
 
 
     """
+    # Unpack kwargs
```
Could remove these.

bohnstingl replied:
Done

bohnstingl committed:
Fixed positional-only arguments for scan

bohnstingl committed:
Fixed lintrunner issues

ydwu4 commented:
@pytorchbot merge

jeffdaily commented:
This PR seems to have broken ROCm CI.

https://hud.pytorch.org/failure?name=rocm%20%2F%20linux-focal-rocm6.1-py3.8%20%2F%20test%20(default%2C%206%2C%206%2C%20linux.rocm.gpu.2)&jobName=undefined&failureCaptures=%5B%22functorch%2Ftest_control_flow.py%3A%3ATestControlFlow%3A%3Atest_associative_scan_binary_operator_combine_mode_pointwise_reverse_False_cuda%22%5D

ClashLuke commented on torch/_higher_order_ops/scan.py:
```diff
@@ -0,0 +1,438 @@
+
+    """
+    if not callable(combine_fn):
+        raise RuntimeError("Combine_fn must be a callable, but got {combine_fn}")
```
should this be an f-string?
